### PR TITLE
Add a simple symbol table.

### PR DESCRIPTION
**Pull request by zrlk**
This PR adds a simple symbol table and the necessary
specailizations for slices and options of symbols.

### CODE REVIEW

File: anodyne/base/symbol_table.h [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+  SymbolTable(const SymbolTable&) = delete;
+  SymbolTable& operator=(const SymbolTable&) = delete;
+  /// \brief Given a non-gensym `Symbol`, return a string that can be used to
+  /// look it up.
+  absl::string_view Text(Symbol symbol) const {
+    assert(symbol < kGensymBase);
+    return symbol_text_[symbol];
+  }
+  /// \return whether `symbol` was generated with `Gensym`.
+  bool is_gensym(Symbol symbol) const { return symbol >= kGensymBase; }
```
Comment (creachadair) [inline anodyne/base/symbol_table.h]: Aw, not using the low-order bit as a tag? :)

Comment (creachadair) [reply anodyne/base/symbol_table.h]: (that wasn't a request for a change, by the way)

File: anodyne/base/symbol_table.h [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+  Symbol Gensym() { return gensym_++; }
+
+ private:
+  std::unordered_map<std::string, Symbol> symbol_map_;
+  std::vector<std::string> symbol_text_;
+  static constexpr uint32_t kGensymBase = 0x80000000;
+  uint32_t gensym_ = kGensymBase;
+};
+
+template <>
```
Comment (creachadair) [inline anodyne/base/symbol_table.h]: Please document these, at least as to intent. I don't entirely follow why they really need to be in this header, given that the symtab doesn't use them.

Comment (zrlk) [reply anodyne/base/symbol_table.h]: They're associated with the `Symbol` type and should be available when that type is available.

**Pull request merged**